src/commonMain/kotlin/CommonJsonLogicEngine.kt

```diff
@@ -1,5 +1,26 @@
+import expressions.Missing
+import expressions.MissingSome
+import expressions.Var
+import expressions.Var.operation
+
 internal class CommonJsonLogicEngine : JsonLogicEngine {
-    override fun evaluate(expression: String, data: Map<String, Any>): String {
-        return "SUCCESS"
+    private val operations: Map<String, (Any?, Any?) -> Any?> = mapOf(
+        Var.operation,
+        MissingSome.operation,
+        Missing.operation
+    )
+
+    override fun evaluate(expression: Map<String, Any?>, data: Any?): Any? = apply(expression, data)
+
+    private fun apply(logic: Any?, data: Any?): Any? {
+        if (logic !is Map<*, *>) return logic
```

Comment:
I think it's better to throw an error here

Reply:
+1, for me the fact that library that I used in PoC returned logic or passed data when I passed incorrect logic or data was confusing

Reply:
Unless there will be another evaluate functions, there is not much possibility to get something other than map here? 🤔

Reply:
Indeed, by the contract we force users to use maps. But inside them there still might be various things 😨

Reply:
There might be another evaluate functions which are part of the recursion. In some cases logic might not be an expression like "operation": "single value" but some lists. For example when `value` is single value its the moment when the recursion calls should stop.